feat: Adds ElementBlock component

chris-decker-volusion opened the pull request:
- Running into a typescript issue on the `defaultConfig` import in `Block.tsx`
- Also getting an `Import of non-existent exports` warning for my types in defined in the `types` file.
- Using placeholders for the `globalSettings` keys as they don't exist yet.
- Haven't bothered with tests until the `globalSettings` keys exist.
- Need to update readme.

Commit:
fix: Adds aphrodite to rollup

Commit:
feat: Adds base files for new component.

chris-decker-volusion committed:
Merge branch 'master' into feat-adds-element-block-component

cameron-jones-volusion commented on src/ElementBlock/.element-block:
```diff
@@ -0,0 +1 @@
+{"activeVersion":1,"displayName":"Elementblock","id":"416084225725302631","idFromStart":true,"publishedName":"ElementBlock"}
```
The `.element-block` file is not used by the component system and can be deleted.

cameron-jones-volusion commented on src/ElementBlock/src/configs.ts:
```diff
@@ -0,0 +1,64 @@
+    }
+};
+
+export const defaultConfig: ElementBlockConfigs = {
+    isFullWidth: false,
+    hasHorizontalPadding: true,
+    topPadding: 'Small',
+    bottomPadding: 'Small',
+    showOnDesktop: true,
+    showOnTablet: true,
+    showOnMobile: true,
+    customAttrs: [
+        {
+            name: '',
+            value: ''
+        }
+    ]
+};
```
For backwards compatibility, the default values should have no visible effect on the wrapped blocks. The result of rendering with the default values should be a wrapping `div` element with no styling or attributes.

cameron-jones-volusion commented on src/ElementBlock/src/getStyles.ts:
```diff
@@ -0,0 +1,65 @@
+    paddingSize: 'None' | 'Small' | 'Medium' | 'Large',
+    globalSettings: ObjectLiteral
+) => {
+    if (/^small$/i.test(paddingSize)) return globalSettings.__________;
+    if (/^medium$/i.test(paddingSize)) return globalSettings.__________;
+    if (/^large$/i.test(paddingSize)) return globalSettings.__________;
```
For now, we can deal with the missing global style variables by setting arbitrary fixed values here.
```suggestion
    if (/^small$/i.test(paddingSize)) return '0.5rem';
    if (/^medium$/i.test(paddingSize)) return '1rem';
    if (/^large$/i.test(paddingSize)) return '1.5rem';    
```

cameron-jones-volusion commented on src/ElementBlock/src/Block.tsx:
```diff
@@ -0,0 +1,33 @@
+            return attr;
+        };
+        const fullAttrName = `data-element-${hyphenateString(attr.name)}`;
+        customProps[fullAttrName] = attr.value;
```
Using `map` & `reduce` here will get rid of the eslint security warning triggered by this line, something like this:
```tsx
    const attrs = customAttrs.map((attr: { name: string; value: string }) => {
        const hyphenateString = (string = '') => {
            const nonAlphaNumericChars = /[^a-zA-Z0-9]+/g;
            const attr = string
                .trim()
                // .replace(specialChars, '')
                .replace(nonAlphaNumericChars, '-');
            return attr;
        };
        return {
            [`data-element-${hyphenateString(attr.name)}`]: value
        }
    })
    .reduce((acc, attr) => {
        return { ...acc, ...attr }
    }, {})
```

chris-decker-volusion replied:
This issue seems to have disappeared...

cameron-jones-volusion reviewed (changes requested):
In order to get this component out, and unblock related work in the Element Platform, we can use static values as placeholders for the variables that we will eventually be getting from the global styles object. That should be enough to write some tests and get this merged.

chris-decker-volusion committed:
Merge branch 'master' into feat-adds-element-block-component

Commit:
chore: adds comments.

Commit:
chore: Deletes element-block file.

Commit:
chore: Updates styles with temp values for globals

Commit:
fix: Updates rollup configs

- Removes duplicate aphrodite from globals.
- Updates elsint to not throw errors on linting.

Commit:
chore: lint reformating.

Commit:
chore: Lints local file.

Commit:
chore: Updates the default configs.

Commit:
feat: Return null if no children.

Commit:
fix: Filter customAttrs for valid names.

Commit:
test: Updates tests to pass and adds new tests.

Commit:
fix: Corrects customAttrsProps type definition.

Commit:
test: Adds new tests

- Updates block handling of special characters in custom attr names

Commit:
feat: Updates custom attrs handling.

Commit:
test: Updates test description.

chris-decker-volusion committed:
Merge branch 'master' into feat-adds-element-block-component

cameron-jones-volusion commented on rollup.config.js:
```diff
@@ -70,7 +72,7 @@ function devBuild() {
         plugins: [
             eslint({
-                throwOnError: true
+                throwOnError: false
```
Useful for local development, but we should remove before merging.

chris-decker-volusion replied:
4ea6b16dfe50d16cf3e9e4a55a38b9eff91fa30e

cameron-jones-volusion commented on src/index.js:
```diff
@@ -10,5 +11,6 @@ export const LegacyComponents = {
     Image,
-    Text
+    Text,
+    ElementBlock
```
This component will not be available through `ElementPropTypes.component()`, so it should not be included in the `StandardComponents` object. For now, we should export it from the top level of the module.

chris-decker-volusion replied:
f98d70e34dab56e1b6221b39cb5bf26014fd78fd

cameron-jones-volusion reviewed (changes requested):
* Only major change is that we should make ElementBlock a top-level export of the package.
* Remove README for now.

Commit:
chore: Deletes readme.

Commit:
fix: Moves ElementBlock to be top-level export.

Commit:
fix: Reverts rollup eslint option change.

chris-decker-volusion committed:
Merge branch 'master' into feat-adds-element-block-component
